packages/recommend-react/src/RecommendContext.tsx

```diff
@@ -0,0 +1,207 @@
+type StateType<TObject> = {
+  isDirty: number | null;
+  cache: Record<string, BatchRecommendations<TObject>>;
+  widgets: Array<RecommendWidget<TObject>>;
+  recommendClient: RecommendClient | null;
+};
+
+type Action<TObject> =
+  | { type: 'register'; widget: Omit<RecommendWidget<TObject>, 'param'> }
+  | { type: 'unregister'; key: string }
+  | { type: 'request_success' };
+
+export const RecommendContext = React.createContext<
+  RecommendContextType<unknown> // Not sure about using "unknown" here ?
+>({
+  hasProvider: false,
+} as RecommendContextType<unknown>);
+
+export const useRecommendContext = () => React.useContext(RecommendContext);
+
+function isRegistered<TObject>(
+  widgets: Array<RecommendWidget<TObject>>,
+  param: GetParametersResult
+) {
+  return Boolean(widgets.find((w) => dequal(w.param, param)));
+}
+
+const getCacheKey = (queries: RecommendationsQuery[]) => {
+  return JSON.stringify(queries);
+};
+
+function isCached<TObject>(
+  cache: StateType<TObject>['cache'],
+  queries: RecommendationsQuery[]
+) {
+  return Boolean(cache[getCacheKey(queries)]);
+}
+
+function getCachedPrams<TObject>(
+  widgets: Array<RecommendWidget<TObject>>,
+  cache: StateType<TObject>['cache']
+) {
+  return widgets
+    .map((widget) => {
+      if (isCached(cache, widget.param.queries)) {
+        const cachedValue = cache[getCacheKey(widget.param.queries)];
+        widget.onResult(cachedValue);
+        return null; // No need to call API if we have value from cache
+      }
+      return widget.param;
+    })
+    .filter(isPresent);
+}
+
+const getQueryKeys = (params: GetParametersResult[]) => {
+  const queries = params
+    .map((p) => p.queries)
+    .reduce((a, b) => a.concat(b), []);
+  const keys = params.map((p) => p.keyPair);
+  return { queries, keys };
+};
+// Not sure about using "unknown"/"any" here ?
+const reducer: React.Reducer<StateType<unknown>, Action<any>> = (
+  state,
+  action
+) => {
+  const { type } = action;
+  switch (type) {
+    case 'register': {
+      const { widget } = action;
+      const param = widget.getParameters();
+      const isPreviouslyRegistered = isRegistered(state.widgets, param);
+      if (!isPreviouslyRegistered) {
+        return {
+          ...state,
+          widgets: [...state.widgets, { ...widget, param }],
+          isDirty: Date.now(),
+        };
+      }
+
+      return state;
+    }
+
+    case 'unregister': {
+      const { key } = action;
+      state.widgets = state.widgets.filter((w) => w.key !== key);
+      return state;
+    }
+
+    case 'request_success': {
+      return { ...state, status: 'idle', isDirty: null };
+    }
+
+    default:
+      return state;
+  }
+};
+
+export function RecommendProvider<TObject>({
+  recommendClient,
+  children,
+}: RecommendProviderProps) {
+  const [state, dispatch] = React.useReducer(reducer, {
+    isDirty: null,
+    cache: {},
+    widgets: [],
+    recommendClient,
+  });
+
+  React.useEffect(() => {
+    (async () => {
```

Comment:
We've been avoiding async functions in our libraries because the tooling doesn't transpile them down to our browser target IIRC.